File: java-spring/src/main/java/org/burgas/javaspring/mapper/NoteMapper.java
package org.burgas.javaspring.mapper;

import lombok.RequiredArgsConstructor;
import org.burgas.javaspring.dto.note.NoteFullResponse;
import org.burgas.javaspring.dto.note.NoteRequest;
import org.burgas.javaspring.dto.note.NoteShortResponse;
import org.burgas.javaspring.entity.note.Note;
import org.burgas.javaspring.repository.NoteRepository;
import org.springframework.beans.factory.ObjectFactory;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Optional;
import java.util.UUID;

@Component
@RequiredArgsConstructor
public final class NoteMapper implements Mapper<NoteRequest, Note, NoteShortResponse, NoteFullResponse> {

    final NoteRepository noteRepository;
    private final ObjectFactory<IdentityMapper> identityMapperObjectFactory;

    private IdentityMapper getIdentityMapper() {
        return this.identityMapperObjectFactory.getObject();
    }

    @Override
    public Note toEntity(NoteRequest request) {
        return this.noteRepository.findById(request.getId() == null ? new UUID(0, 0) : request.getId())
                .map(
                        note -> {
                            var identityId = Optional.ofNullable(request.getId()).orElse(new UUID(0, 0));
                            var findIdentity = this.getIdentityMapper().identityRepository.findById(identityId)
                                    .orElse(null);
                            var identity = Optional.ofNullable(findIdentity).orElse(note.getIdentity());
                            return Note.builder()
                                    .id(note.getId())
                                    .title(Optional.ofNullable(request.getTitle()).orElse(note.getTitle()))
                                    .content(Optional.ofNullable(request.getContent()).orElse(note.getContent()))
                                    .createdAt(note.getCreatedAt())
                                    .identity(identity)
                                    .build();
                        }
                )
                .orElseGet(
                        () -> {
                            var identityId = Optional.ofNullable(request.getIdentityId()).orElseThrow();
                            var findIdentity = this.getIdentityMapper().identityRepository.findById(identityId)
                                    .orElse(null);
                            var identity = Optional.ofNullable(findIdentity).orElseThrow();
                            return Note.builder()
                                    .title(Optional.ofNullable(request.getTitle()).orElseThrow())
                                    .content(Optional.ofNullable(request.getContent()).orElseThrow())
                                    .createdAt(LocalDateTime.now())
                                    .identity(identity)
                                    .build();
                        }
                );
    }

    @Override
    public NoteShortResponse toShortResponse(Note entity) {
        return NoteShortResponse.builder()
                .id(entity.getId())
                .title(entity.getTitle())
                .content(entity.getContent())
                .createdAt(entity.getCreatedAt().format(DateTimeFormatter.ofPattern("dd MMMM yyyy, hh:mm")))
                .images(entity.getImages())
                .build();
    }

    @Override
    public NoteFullResponse toFullResponse(Note entity) {
        return NoteFullResponse.builder()
                .id(entity.getId())
                .title(entity.getTitle())
                .content(entity.getContent())
                .createdAt(entity.getCreatedAt().format(DateTimeFormatter.ofPattern("dd MMMM yyyy, hh:mm")))
                .identity(
                        Optional.ofNullable(entity.getIdentity())
                                .map(identity -> this.getIdentityMapper().toShortResponse(identity)).orElse(null)
                )
                .images(entity.getImages())
                .build();
    }
}

File: kotlin-spring/src/main/kotlin/org/burgas/kotlinspring/mapper/NoteMapper.kt
package org.burgas.kotlinspring.mapper

import org.burgas.kotlinspring.entity.note.Note
import org.burgas.kotlinspring.entity.note.NoteFullResponse
import org.burgas.kotlinspring.entity.note.NoteRequest
import org.burgas.kotlinspring.entity.note.NoteShortResponse
import org.burgas.kotlinspring.repository.NoteRepository
import org.springframework.beans.factory.ObjectFactory
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.Optional
import java.util.UUID

@Component
class NoteMapper : Mapper<NoteRequest, Note, NoteShortResponse, NoteFullResponse> {

    final val noteRepository: NoteRepository
    private final val identityMapperObjectFactory: ObjectFactory<IdentityMapper>

    constructor(noteRepository: NoteRepository, identityMapperObjectFactory: ObjectFactory<IdentityMapper>) {
        this.noteRepository = noteRepository
        this.identityMapperObjectFactory = identityMapperObjectFactory
    }

    private fun getIdentityMapper(): IdentityMapper = this.identityMapperObjectFactory.`object`

    override fun toEntity(request: NoteRequest): Note {
        return this.noteRepository.findById(request.id ?: UUID(0, 0))
            .map {
                Note().apply {
                    this.id = it.id
                    this.title = request.title ?: it.title
                    this.content = request.content ?: it.content
                    this.createdAt = it.createdAt
                    this.identity = getIdentityMapper().identityRepository.findById(request.identityId ?: UUID(0, 0))
                        .orElse(null) ?: it.identity
                }
            }
            .orElseGet {
                Note().apply {
                    this.title = request.title ?: throw IllegalArgumentException("Note title is null")
                    this.content = request.content ?: throw IllegalArgumentException("Note content is null")
                    this.createdAt = LocalDateTime.now()
                    this.identity = getIdentityMapper().identityRepository.findById(request.identityId ?: UUID(0, 0))
                        .orElse(null) ?: throw IllegalArgumentException("Note identity is null")
                }
            }
    }

    override fun toShortResponse(entity: Note): NoteShortResponse {
        return NoteShortResponse(
            id = entity.id,
            title = entity.title,
            content = entity.content,
            createdAt = entity.createdAt.format(DateTimeFormatter.ofPattern("dd MMMM yyyy")),
            images = entity.images
        )
    }

    override fun toFullResponse(entity: Note): NoteFullResponse {
        return NoteFullResponse(
            id = entity.id,
            title = entity.title,
            content = entity.content,
            createdAt = entity.createdAt.format(DateTimeFormatter.ofPattern("dd MMMM yyyy")),
            identity = Optional.ofNullable(entity.identity).map { this.getIdentityMapper().toShortResponse(it) }
                .orElse(null),
            images = entity.images
        )
    }
}